File: Spiral_matrix.cpp
class Solution {
	public:
		vector<int> spiralOrder(vector<vector<int> > &matrix) {
			vector<int> res;

			if(matrix.size() > 0)
				printmatrix(res, matrix, 0, matrix.size()-1, matrix[0].size()-1);

			return res;
		}

		void printmatrix(vector<int> &res, vector<vector<int> >&matrix, int begin, int m, int n)
		{
			if((begin >= m) || (begin >= n))
			{
				if((begin == m) && (m == n))
					res.push_back(matrix[begin][begin]);
				else if((begin == m) && (n > begin))
				{
					for(int i = begin; i <= n; i++)
						res.push_back(matrix[begin][i]);
				}
				else if((begin == n) && (m > begin))
				{
					for(int i = begin; i <= m; i++)
						res.push_back(matrix[i][begin]);
				}
				return ;
			}
			else
			{
				//row
				for(int i = begin; i < n; i++)
				{
					res.push_back(matrix[begin][i]);
				}

				//col
				for(int i = begin; i < m; i++)
				{
					res.push_back(matrix[i][n]);
				}

				//bottom row

				for(int i = n; i > begin; i--)
				{
					res.push_back(matrix[m][i]);
				}

				//left col
				for(int i = m; i > begin; i--)
				{
					res.push_back(matrix[i][begin]);
				}

				printmatrix(res, matrix, begin+1, m-1, n-1);
			}
		}
};

File: leetcode/array/Spiral Matrix.go
func spiralOrder(matrix [][]int) []int {
    res := []int{}
    if len(matrix) == 0{
        return res
    }
    helper(matrix, 0, 0, len(matrix)-1, len(matrix[0])-1, &res)
    return res
}

func helper(matrix [][]int, r1 int, c1 int,  r2 int, c2 int, res *[]int){
    if r2 < r1 || c2 < c1{
        return 
    }
    if r1 == r2{
        for i := c1; i <= c2; i++{
            *res = append(*res, matrix[r1][i])
        }
        return 
    }
    if c1 == c2{
        for i := r1; i <= r2; i++{
            *res = append(*res, matrix[i][c1])
        }
        return 
    }
    for i := c1; i < c2; i++{
        *res = append(*res, matrix[r1][i])
    }

    for i := r1; i < r2; i++{
        *res = append(*res, matrix[i][c2])
    }

    for i := c2; i > c1; i--{
        *res = append(*res, matrix[r2][i])
    }

    for i := r2; i > r1; i--{
        *res = append(*res, matrix[i][c1])
    }

    helper(matrix, r1+1, c1+1, r2-1, c2-1, res)
}